Media & text: Update the image replacement logic

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Updates the logic that inserts the featured image in the media and text block, taking in to account if the media is set to be on the left or right side. 

For https://github.com/WordPress/gutenberg/issues/62023

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
When the media is on the right, and there is a second media & text block inside the block, the featured image was inserted inside the wrong HTML element.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
First, the updated code checks if the block's `mediaPosition` attribute is set to `right`.
If it is, it locates the **last** figure inside the Media & text block, and inserts an `img` tag inside it.
If not, the `img` tag  is inserted inside the first  figure.

It identifies the correct figure element by the class wp-block-media-text__media to not get mixed up with other figure elements.
It identifies the inserted img tag with the class wp-block-media-text__featured_image.

Then the attributes on the img tag are updated with the data for the featured image, such as the `src`.

## Testing Instructions
Apply the PR.
Create a new post and add a featured image.
Insert a media & text block.
Enable using the featured image option on the block.
Enable the option to position the media on the right.

Now insert another media & text block in the first media & text block. It doesn't matter if this block has any content or has a media selected.
View the front of the website, and compare the size and position of the featured image in the editor and front.


Without this PR; the HTML will look something like this on the front.
Note that there are two img elements inside the first figure.

```
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile has-primary-background-color has-background">
<div class="wp-block-media-text__content">
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile is-vertically-aligned-bottom">
<div class="wp-block-media-text__content"></div>
<figure class="wp-block-media-text__media">
<img decoding="async" width="282" height="159" alt="" class="wp-image-240 size-full" src="...">
<img fetchpriority="high" decoding="async" width="775" height="1024" src="..." alt="" class="wp-image-30 size-full" srcset="..." sizes="(max-width: 775px) 100vw, 775px">
</figure>
</div>
</div>
<figure class="wp-block-media-text__media"><img></figure>
</div>
```

And with the PR:

```
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile has-primary-background-color has-background">
<div class="wp-block-media-text__content">
<div class="wp-block-media-text has-media-on-the-right is-stacked-on-mobile is-vertically-aligned-bottom">
<div class="wp-block-media-text__content"></div>
<figure class="wp-block-media-text__media">
<img fetchpriority="high" decoding="async" width="775" height="1024" src="..." alt="" class="wp-image-30 size-full" srcset="..." sizes="(max-width: 775px) 100vw, 775px"></figure>
</div>
</div>
<figure class="wp-block-media-text__media"><img decoding="async" width="282" height="159" alt="" src="..." class="wp-image-240 size-full"></figure>
</div>

```

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

carolinan committed:
Media & text: Update the image replacement logic when the media is on the right

carolinan committed:
Try to fix PHP coding standard issues.

carolinan committed:
Look for the figure with the correct class name instead of any figure  tag.

carolinan commented:
Perhaps a unique identifier needs to be added to the figure that contains the media, that can then be used in the PHP file, but I'm really not sure.

dmsnell commented:
> If anyone is available, I need help with using the HTML tag processor to locate the last `<figure class="wp-block-media-text__media">` in the media and text block when the mediaPosition attribute is set to right.

It's easy enough to find the first and last figure with this class name, but the HTML API doesn't yet support inserting a node.

We _can_ get around this by subclassing, but so far we've been reluctant to add workaround code in Core/Gutenberg so as to be careful not to share the wrong example code and accidentally encourage dangerous behavior.

For instance, will this always inject an `IMG` element? What if the `FIGURE` already contains an `IMG`?

The general approach for finding the first is easy, and for finding the last there's a kind of bookmark crawl to take.

```php
$media_tag_processor   = new WP_HTML_Tag_Processor( $content );
$wrapping_figure_query = array( 'tag_name' => 'figure', 'class_name' => 'wp-block-media-text__media' );
if ( $has_media_on_right ) {
	while ( $media_tag_processor->next_tag( $wrapping_figure_query ) ) {
		$media_tag_processor->set_bookmark( 'last_figure' );
	}
	
	if ( $media_tag_processor->has_bookmark( 'last_figure' ) ) {
		// Modify last figure here.
	}
} else {
	if ( $media_tag_processor->next_tag( $wrapping_figure_query ) ) {
		// Modify first figure here.
	}
}

$content = $media_tag_processor->get_updated_html();
```

The HTML API tries to avoid handing out string indices, and while design work is still ongoing for insertion/removal operations, there's no built-in support. For subclasses, there's a way to achieve this through bookmarking.

```php
$media_tag_processor = new class( $content ) extends WP_HTML_Tag_Processor {
	public function insert_after( $dangerous_raw_html_reject_this_patch ) {
		if (
			WP_HTML_Tag_Processor::STATE_READY === $this->parser_state ||
			WP_HTML_Tag_Processor::STATE_INCOMPLETE_INPUT === $this->parser_state ||
			WP_HTML_Tag_Processor::STATE_COMPLETE === $this->parser_state
		) {
			return false;
		}

		$this->set_bookmark( 'here' );
		$here = $this->bookmarks['here'];
		$this->lexical_updates[] = new WP_HTML_Text_Replacement(
			$here->start,
			$here->length,
			$dangerous_raw_html_reject_this_patch
		);
	}
};
```

Inserting an `IMG` tag is largely benign.

carolinan committed:
Merge branch 'trunk' into try/media-text--media-on-right

carolinan committed:
Update the logic for inserting the featured image

Use ´while´ to locate the correct ´figure´ tag depending on the block settings.
Add support for the image fill setting.

carolinan committed:
Try to fix PHP CS issues with spacing and the arrays.

carolinan committed:
Try to fix PHP CS issues with spacing inside the arrays.

carolinan commented:
@dmsnell  Thank you. @sirreal suggested something similar.

With this update I am not extending the class but still using  `preg_replace` to insert the image tag.
I still need someone to review the logic and the code quality with critical eyes.
But also to consider if there are any other scenarios to test where the feature can break, besides adding a media & text block  or image block inside a media & text block.

ramonjd commented:
I didn't have a lot of time sorry, but I did a smoke test locally and the frontend appears as I'd expect, even with multiple nested media & text blocks. 🎉 

| Before | After |
|--------|--------|
|<img width="400" alt="Screenshot 2024-05-29 at 10 44 24 am" src="https://github.com/WordPress/gutenberg/assets/6458278/0a108bdd-7f44-42e1-9b6c-db1516ce183a"> | <img width="400" alt="Screenshot 2024-05-29 at 10 44 36 am" src="https://github.com/WordPress/gutenberg/assets/6458278/1d1c684d-867f-4863-ac74-6213704ffaac"> | 


There was one bug I found with the link but I see there's an [issue](https://github.com/WordPress/gutenberg/issues/60497) for it already.

Given the constraints I think your approach is fine - the use of the unique id seems like a clever way to target the figure element, and `preg_replace` does the job.

> But also to consider if there are any other scenarios to test where the feature can break, besides adding a media & text block or image block inside a media & text block.

It might help other folks to understand the intention behind the code, and also to have confidence that it doesn't break in various scenarios by getting some test coverage in there, focussing on the expected output. 

See other tests for block render functions:

https://github.com/WordPress/gutenberg/tree/7988696ec25a7373337d91a359929aad204307fd/phpunit/blocks

carolinan commented on packages/block-library/src/media-text/index.php:
```diff
@@ -29,28 +29,76 @@ function render_block_core_media_text( $attributes, $content ) {
+		);
+
+		$image_tag_processor = new WP_HTML_Tag_Processor( $content );
+		$image_tag_processor->next_tag(
```
Maybe this should be "_if_ next tag, do this".

ramonjd replied:
Yeah, good point. So something like

```php
		if ( $image_tag_processor->next_tag(
			array(
				'tag_name' => 'figure',
				'id'       => $unique_id,
			)
		) ) {
			// The ID is only used to ensure that the correct figure tag is selected,
			// and can now be removed.
			$image_tag_processor->remove_attribute( 'id' );
			$image_tag_processor->next_tag(
				array(
					'tag_name'   => 'img',
					'class_name' => 'wp-block-media-text__featured_image',
				)
			);
			$image_tag_processor->set_attribute( 'src', esc_url( $current_featured_image ) );
			$image_tag_processor->set_attribute( 'class', 'wp-image-' . get_post_thumbnail_id() . ' size-' . $media_size_slug );
			$image_tag_processor->set_attribute( 'alt', trim( strip_tags( get_post_meta( get_post_thumbnail_id(), '_wp_attachment_image_alt', true ) ) ) );

			$content = $image_tag_processor->get_updated_html();
		}
```

ramonjd replied:
Also maybe for the `tag_name' => 'img',` one as well.

carolinan commented on packages/block-library/src/media-text/index.php:
```diff
@@ -29,28 +29,76 @@ function render_block_core_media_text( $attributes, $content ) {
+	$focal_point           = isset( $attributes['focalPoint'] ) ? round( $attributes['focalPoint']['x'] * 100 ) . '% ' . round( $attributes['focalPoint']['y'] * 100 ) . '%' : '50% 50%';
+	$media_size_slug       = isset( $attributes['mediaSizeSlug'] ) ? $attributes['mediaSizeSlug'] : 'full';
+	$unique_id             = 'wp-block-media-text__media-' . wp_unique_id();
+	$image_tag             = '<img class="wp-block-media-text__featured_image">';
```
This class name is probably not needed, and does not exist in the editor, which makes it difficult to use it for styling anyway.

carolinan replied:
I forgot that I already removed it from the resulting markup, it is _only_ used for identifying the image tag.

ramonjd replied:
Do you need it at all?

Wouldn't we expect an image to come after the figure? E.g., 

```php
		$image_tag_processor->next_tag(
			array(
				'tag_name'   => 'img',
		);
```

Otherwise you could delete it as you do with the id. Probably best to remove if it's not required 👍🏻

carolinan committed:
Add a test class for the Media & Text block

carolinan commented:
> It might help other folks to understand the intention behind the code, and also to have confidence that it doesn't break in various scenarios by getting some test coverage in there, focussing on the expected output.
> 
> See other tests for block render functions:
> 
> https://github.com/WordPress/gutenberg/tree/7988696ec25a7373337d91a359929aad204307fd/phpunit/blocks

I have tried, but I have never written a test class before. Maybe it should be 3 separate tests, if that is easier to read? One for default settings, one for image fill, and one for when the media is on the right. Now they are all in the same test.

commands:
`npm run test:unit:php  `

Or (after starting wp-env)
`wp-env run --env-cwd='wp-content/plugins/gutenberg' tests-wordpress vendor/bin/phpunit -c phpunit.xml.dist --verbose --filter Render_Block_MediaText_Test`

carolinan committed:
Merge branch 'trunk' into try/media-text--media-on-right

ramonjd commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -0,0 +1,128 @@
+			'mediaPosition'    => 'right',
+		);
+		$rendered   = gutenberg_render_block_core_media_text( $attributes, $content_media_on_right );
+		$this->assertStringContainsString( wp_get_attachment_image_url( self::$attachment_id, 'full' ), $rendered );
```
```suggestion
		$this->assertStringContainsString( '<img alt="" src="' . wp_get_attachment_image_url( self::$attachment_id, 'full' ) . '"', $rendered );
```

Thanks a lot for the tests! I know these are tricky to write, especially where unique ids are being generated. 

Here we could check that the image is at least part of an IMG element I think. There might be one or two above I missed.

ramonjd commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -0,0 +1,128 @@
+		$this->assertStringContainsString( wp_get_attachment_image_url( self::$attachment_id, 'full' ), $rendered );
+
+		$rendered = gutenberg_render_block_core_media_text( $attributes, $content_nested_media_on_right );
+		$this->assertStringContainsString( wp_get_attachment_image_url( self::$attachment_id, 'full' ), $rendered );
```
```suggestion
		$this->assertStringContainsString( '<img alt="" src="' . wp_get_attachment_image_url( self::$attachment_id, 'full' ) . '"', $rendered );
```

This one too.

ramonjd commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -0,0 +1,128 @@
+ *
+ * @covers ::render_block_core_media_text
+ */
+//class Tests_Blocks_Render_MediaText extends WP_UnitTestCase {
```
Tiny nit - can remove

carolinan replied:
Oh yeah I was trying to figure out what the best practice was for the naming 😺

ramonjd commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -0,0 +1,128 @@
+			'useFeaturedImage' => true,
+		);
+		$rendered   = gutenberg_render_block_core_media_text( $attributes, $content );
+		$this->assertStringContainsString( wp_get_attachment_image_url( self::$attachment_id, 'full' ), $rendered );
```
> I have tried, but I have never written a test class before. Maybe it should be 3 separate tests, if that is easier to read? One for default settings, one for image fill, and one for when the media is on the right. Now they are all in the same test.

It's pretty good! Great job.

I would split them up into 3 tests as they'll be easier to read and maintain.

At the very least I'd add the third message argument to the assertion methods so it's easier to know which test fails, e.g., 

```php
$this->assertStringContainsString( wp_get_attachment_image_url( self::$attachment_id, 'full' ), $rendered, 'The rendered block does not contain a featured image blah blah' );
```

Alternatively you could use a data provider ([Example](https://github.com/WordPress/gutenberg/blob/ee92f208805462a13229a4e4d5baee6898dec734/phpunit/block-supports/layout-test.php#L413-L413)) to loop over a series of test arguments.

But I think just splitting them up into separate functions would be find if you feel comfortable doing that.

ramonjd reviewed (commented):
Thanks for the continued work on this! And also for the tests.

I think it'd be good to break the tests up for readability and also to help isolate fails.

carolinan committed:
Replace the $image_tag variable and remove the CSS class on the img tag

Neither the variable or the CSS class are needed, and they can be removed in favor of an empty
`<img>` tag.

carolinan commented:
Actually removing the class name from the img did break the nested media & text block so I need to revert it.

carolinan committed:
Revert "Replace the $image_tag variable and remove the CSS class on the img tag"

This reverts commit c62c46cc966bd59103463335a45330f9ebc335e9.

carolinan committed:
Add an if statement around the figure tag and the image tag

Add an if statement around the figure tag and the image tag, to make sure that the id is removed and the attributes are updated on the correct tags.

carolinan committed:
Try to fix PHP CS issues

Move the closing parenthesis on the array that includes the figure tag name and id.
Remove whitespace.

carolinan committed:
Split test_render_block_core_media_text_featured_image into four tests

carolinan commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -94,21 +87,66 @@ public function test_render_block_core_media_text_featured_image() {
+		$wp_query->post        = self::$post;
+
+		$wp_query->posts = array( self::$post );
+		$GLOBALS['post'] = self::$post;
```
I want to try to see if I can avoid repeating this code but I might not have time until the end of June.

ramonjd replied:
There are only three tests, so I reckon it's fine to iterate. 

Thanks a lot for splitting them up!

ramonjd commented on packages/block-library/src/media-text/index.php:
```diff
@@ -29,28 +29,78 @@ function render_block_core_media_text( $attributes, $content ) {
+	$has_media_on_right    = isset( $attributes['mediaPosition'] ) && 'right' === $attributes['mediaPosition'];
+	$image_fill            = isset( $attributes['imageFill'] ) && $attributes['imageFill'];
+	$focal_point           = isset( $attributes['focalPoint'] ) ? round( $attributes['focalPoint']['x'] * 100 ) . '% ' . round( $attributes['focalPoint']['y'] * 100 ) . '%' : '50% 50%';
+	$media_size_slug       = isset( $attributes['mediaSizeSlug'] ) ? $attributes['mediaSizeSlug'] : 'full';
```
Could this be moved down, immediately before it's used?

ramonjd commented on packages/block-library/src/media-text/index.php:
```diff
@@ -29,28 +29,78 @@ function render_block_core_media_text( $attributes, $content ) {
+	$focal_point           = isset( $attributes['focalPoint'] ) ? round( $attributes['focalPoint']['x'] * 100 ) . '% ' . round( $attributes['focalPoint']['y'] * 100 ) . '%' : '50% 50%';
+	$media_size_slug       = isset( $attributes['mediaSizeSlug'] ) ? $attributes['mediaSizeSlug'] : 'full';
+	$unique_id             = 'wp-block-media-text__media-' . wp_unique_id();
+	$image_tag             = '<img class="wp-block-media-text__featured_image">';
```
Same here. Could this be moved inside the `if ( ! $image_fill ) {` block below?

ramonjd commented on packages/block-library/src/media-text/index.php:
```diff
@@ -29,28 +29,78 @@ function render_block_core_media_text( $attributes, $content ) {
+			) ) {
+				$image_tag_processor->set_attribute( 'src', esc_url( $current_featured_image ) );
+				$image_tag_processor->set_attribute( 'class', 'wp-image-' . get_post_thumbnail_id() . ' size-' . $media_size_slug );
+				$image_tag_processor->set_attribute( 'alt', trim( strip_tags( get_post_meta( get_post_thumbnail_id(), '_wp_attachment_image_alt', true ) ) ) );
```
Idea for possible follow up: When the image is a background image add accessible alt text via `aria-label` (or whatever's appropriate) on the figure element

ramonjd commented on phpunit/blocks/render-block-media-text-test.php:
```diff
@@ -0,0 +1,201 @@
+	/**
+	 * Post object.
+	 *
+	 * @var object
```
```suggestion
	 * @var WP_Post
```

ramonjd reviewed (approved):
I've tested manually with single media and text and nested. I think this is an improvement 👍🏻 

Thanks for sticking with it.

| Before  | After |
| ------------- | ------------- |
| <img width="400" alt="Screenshot 2024-06-04 at 12 04 13 PM" src="https://github.com/WordPress/gutenberg/assets/6458278/468ece29-9761-4a5f-9989-df7fbfdc0f3a"> | <img width="400" alt="Screenshot 2024-06-04 at 11 59 37 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/34755419-9f6c-4bb2-bba0-ec36de35951e">  |


If @dmsnell has time, it'd be good to get his keen eye to run over the HTML Processor changes. 



I noticed a bug - a nested media and text block with an alignment of left does not align left in the frontend or the editor. 

<img width="400" alt="Screenshot 2024-06-04 at 11 52 53 AM" src="https://github.com/WordPress/gutenberg/assets/6458278/6e5602e2-ec5d-4154-80e2-2c8687ecc78a">

I think this is an existing bug however.

carolinan committed:
Update phpunit/blocks/render-block-media-text-test.php

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

carolinan commented:
> I noticed a bug - a nested media and text block with an alignment of left does not align left in the frontend or the editor.
This should be fixed: https://github.com/WordPress/gutenberg/pull/62184

carolinan committed:
Merge branch 'trunk' into try/media-text--media-on-right

carolinan committed:
Move the PHP variables for media size slug and image tag

carolinan committed:
Try to fix PHPCS spacing issues

dmsnell commented:
> If @dmsnell has time, it'd be good to get his keen eye to run over the HTML Processor changes.

It'd be nice to get rid of the regex, but that's not convenient or easy with the current limits of the HTML API. I'm happy with how the code is because at least the search for the right place to add the classes is robust. IMO it's easier to clean up existing code patterns (the `regex_replace`) than it is to start hunting down custom subclasses of the HTML API that extend its functionality.

In other words, I still think this is appropriate w.r.t. the HTML processing.

carolinan committed:
Merge branch 'trunk' into try/media-text--media-on-right

carolinan committed:
Add a helper method for wp_query (post data) to reduce code duplication

carolinan committed:
PHP coding standards: adjust spacing around equal signs.

ellatrix commented:
I just cherry-picked this PR to the wp/6.6-rc-1 branch to get it included in the next release: ab687f7939
